google-beta/resource_pubsub_topic.go

```diff
@@ -39,74 +65,145 @@ func resourcePubsubTopic() *schema.Resource {
-	res, err := call.Do()
+	res, err := sendRequest(config, "GET", url, nil)
+	if err != nil {
+		return handleNotFoundError(err, d, fmt.Sprintf("PubsubTopic %q", d.Id()))
+	}
+
+	project, err := getProject(d, config)
 	if err != nil {
-		return handleNotFoundError(err, d, fmt.Sprintf("Pubsub Topic %q", name))
+		return err
+	}
+	if err := d.Set("project", project); err != nil {
+		return fmt.Errorf("Error reading Topic: %s", err)
 	}
 
-	d.Set("name", GetResourceNameFromSelfLink(res.Name))
-	d.Set("project", project)
+	if err := d.Set("name", flattenPubsubTopicName(res["name"], d)); err != nil {
+		return fmt.Errorf("Error reading Topic: %s", err)
+	}
+	if err := d.Set("labels", flattenPubsubTopicLabels(res["labels"], d)); err != nil {
+		return fmt.Errorf("Error reading Topic: %s", err)
+	}
 
 	return nil
 }
 
 func resourcePubsubTopicDelete(d *schema.ResourceData, meta interface{}) error {
 	config := meta.(*Config)
 
-	name := d.Id()
-	call := config.clientPubsub.Projects.Topics.Delete(name)
-	_, err := call.Do()
+	url, err := replaceVars(d, config, "https://pubsub.googleapis.com/v1/projects/{{project}}/topics/{{name}}")
 	if err != nil {
 		return err
 	}
 
+	var obj map[string]interface{}
+	log.Printf("[DEBUG] Deleting Topic %q", d.Id())
+	res, err := sendRequestWithTimeout(config, "DELETE", url, obj, d.Timeout(schema.TimeoutDelete))
+	if err != nil {
+		return handleNotFoundError(err, d, "Topic")
+	}
+
+	log.Printf("[DEBUG] Finished deleting Topic %q: %#v", d.Id(), res)
 	return nil
 }
 
-func resourcePubsubTopicStateImporter(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+func resourcePubsubTopicImport(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
 	config := meta.(*Config)
+	if err := parseImportId([]string{"projects/(?P<project>[^/]+)/topics/(?P<name>[^/]+)", "(?P<project>[^/]+)/(?P<name>[^/]+)", "(?P<name>[^/]+)"}, d, config); err != nil {
+		return nil, err
+	}
 
-	topicId := regexp.MustCompile("^projects/[^/]+/topics/[^/]+$")
-	if topicId.MatchString(d.Id()) {
-		return []*schema.ResourceData{d}, nil
+	// Replace import id for the resource id
+	id, err := replaceVars(d, config, "projects/{{project}}/topics/{{name}}")
+	if err != nil {
+		return nil, fmt.Errorf("Error constructing id: %s", err)
 	}
+	d.SetId(id)
+
+	return []*schema.ResourceData{d}, nil
+}
 
-	if config.Project == "" {
-		return nil, fmt.Errorf("The default project for the provider must be set when using the `{name}` id format.")
+func flattenPubsubTopicName(v interface{}, d *schema.ResourceData) interface{} {
+	if v == nil {
+		return v
 	}
+	return NameFromSelfLinkStateFunc(v)
+}
 
-	id := fmt.Sprintf("projects/%s/topics/%s", config.Project, d.Id())
+func flattenPubsubTopicLabels(v interface{}, d *schema.ResourceData) interface{} {
+	return v
+}
 
-	d.SetId(id)
+func expandPubsubTopicName(v interface{}, d *schema.ResourceData, config *Config) (interface{}, error) {
```

Comment:
Same as subscription, replacevars expects the short name; we can use `https://github.com/GoogleCloudPlatform/magic-modules/blob/master/templates/terraform/custom_expand/resource_from_self_link.go.erb`

Reply:
It looks like we don't need to preserve existing behaviour here; `Create` actually relied on a short name being provided & a long name would fail.

Reply:
ACK